pageserver/src/tenant.rs

```diff
@@ -1231,9 +1235,51 @@ impl Tenant {
 
     pub async fn apply_timeline_archival_config(
         &self,
-        _timeline_id: TimelineId,
-        _config: TimelineArchivalState,
+        timeline_id: TimelineId,
+        state: TimelineArchivalState,
     ) -> anyhow::Result<()> {
+        let timeline = self
+            .get_timeline(timeline_id, false)
+            .context("Cannot apply timeline archival config to inexistent timeline")?;
+
+        let is_archived = timeline.is_archived();
+
+        let upload_needed = match (is_archived, state) {
+            (true, TimelineArchivalState::Archived)
+            | (false, TimelineArchivalState::Unarchived) => {
+                // Nothing to do
+                tracing::info!("intended state matches present state");
+                None
+            }
+            (false, TimelineArchivalState::Archived) => Some(Some(Utc::now().naive_utc())),
+            (true, TimelineArchivalState::Unarchived) => Some(None),
+        };
+
+        if let Some(intended_archived_at) = upload_needed {
+            // We have to schedule this into the upload queue with the other operations,
+            // as if we side-step the upload queue and do it immediately, later operations
+            // without the `archived_at` field might override the file.
+            // Deletion addresses this by shutting down the upload queue. We cannot afford
+            // this here.
+            timeline
+                .remote_client
+                .schedule_index_upload_for_archived_at_update(intended_archived_at)?;
```

Comment:
consider using tokio::timeout https://docs.rs/tokio/latest/tokio/time/fn.timeout.html to simplify the code?

Reply:
done